src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3941,38 +3962,33 @@ public void EnsureInitialized()
         /// </summary>
         [DebuggerStepThrough]
         [DebuggerHidden]
-        internal object? CreateInstanceDefaultCtor(bool publicOnly, bool skipCheckThis, bool fillCache, bool wrapExceptions)
+        internal unsafe object? CreateInstanceDefaultCtor(bool publicOnly, bool skipCheckThis, bool fillCache, bool wrapExceptions)
         {
-            // Call the cached
-            if (GenericCache is ActivatorCache cacheEntry)
+            // Call the cached factory if it exists
+            if (GenericCache is ActivatorCache cache)
             {
-                cacheEntry.EnsureInitialized();
+                cache.EnsureInitialized(this);
+
+                Debug.Assert(cache._pfnNewobj != null);
+                Debug.Assert(cache._pfnCtor != null);
 
-                if (publicOnly)
+                if (!cache._ctorIsPublic && publicOnly)
                 {
-                    if (cacheEntry._ctor != null &&
-                        (cacheEntry._ctorAttributes & MethodAttributes.MemberAccessMask) != MethodAttributes.Public)
-                    {
-                        throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, this));
-                    }
+                    throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, this));
                 }
 
-                // Allocate empty object and call the default constructor if present.
-                object instance = RuntimeTypeHandle.Allocate(this);
```

Comment:
`RuntimeTypeHandle.Allocate` can be deleted now. The two remaining uses can be switched to `RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter`.

Reply:
You can ignore this comment. This can be separate cleanup. `RuntimeTypeHandle.Allocate` is used to create instantiations with more than one generic argument, so it is not easy to replace with `CreateInstanceForAnotherGenericParameter`.